kmip_mgmt_adapters.go

```diff
@@ -0,0 +1,157 @@
+package kp
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"strconv"
+	"time"
+)
+
+const (
+	KMIPAdapterPath = "kmip_adapters"
+	kmipAdapterType = "application/vnd.ibm.kms.kmip_adapter+json"
+)
+
+type KMIPAdapter struct {
+	ID          string      `json:"id,omitempty"`
+	Profile     string      `json:"profile,omitempty"`
+	ProfileData kmipProfile `json:"profile_data,omitempty"`
+	Name        string      `json:"name,omitempty"`
+	Description string      `json:"description"`
+	CreatedBy   string      `json:"created_by,omitempty"`
+	CreatedAt   *time.Time  `json:"created_at,omitempty"`
+	UpdatedBy   string      `json:"updated_by,omitempty"`
+	UpdatedAt   *time.Time  `json:"updated_at,omitempty"`
+}
+
+type KMIPAdapters struct {
+	Metadata KeysMetadata  `json:"metadata"`
+	Adapters []KMIPAdapter `json:"resources"`
+}
+
+type kmipProfile interface {
+	Type() string
+}
+
+type KMIPProfileNative struct {
+	CrkID string `json:"crk_id"`
+}
+
+func (k KMIPProfileNative) Type() string { return KMIP_Profile_Native }
+
+const (
+	KMIP_Profile_Native = "native_1.0"
+)
+
+// CreateKMIPAdapter method creates a KMIP Adapter with the specified profile.
+// For information please refer to the link below:
+// https://cloud.ibm.com/docs/key-protect?topic=placeholder
+func (c *Client) CreateKMIPAdapter(ctx context.Context, profile CreateKMIPAdapterProfile, options ...CreateKMIPAdapterOption) (*KMIPAdapter, error) {
+	newAdapter := &KMIPAdapter{}
+	profile(newAdapter)
+	for _, opt := range options {
+		opt(newAdapter)
+	}
+	req, err := c.newRequest("POST", KMIPAdapterPath, wrapKMIPAdapter(*newAdapter))
+	if err != nil {
+		return nil, err
+	}
+
+	create_resp := &KMIPAdapters{}
+	_, err = c.do(ctx, req, create_resp)
+	if err != nil {
+		return nil, err
+	}
+
+	return unwrapKMIPAdapterResp(create_resp), nil
+}
+
+type CreateKMIPAdapterOption func(*KMIPAdapter)
+type CreateKMIPAdapterProfile func(*KMIPAdapter)
+
+func WithKMIPAdapterName(name string) CreateKMIPAdapterOption {
+	return func(adapter *KMIPAdapter) {
+		adapter.Name = name
+	}
+}
+
+func WithKMIPAdapterDescription(description string) CreateKMIPAdapterOption {
+	return func(adapter *KMIPAdapter) {
+		adapter.Description = description
+	}
+}
+
+func WithNativeProfile(crkID string) CreateKMIPAdapterProfile {
+	return func(adapter *KMIPAdapter) {
+		adapter.Profile = KMIP_Profile_Native
+
+		adapter.ProfileData = KMIPProfileNative{
+			CrkID: crkID,
+		}
+	}
+}
+
+func (c *Client) GetKMIPAdapters(ctx context.Context, limit, offset int) (*KMIPAdapters, error) {
+	adapters := KMIPAdapters{}
+	req, err := c.newRequest("GET", KMIPAdapterPath, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	v := url.Values{}
+	v.Set("limit", strconv.Itoa(limit))
+	v.Set("offset", strconv.Itoa(offset))
+	req.URL.RawQuery = v.Encode()
```

Comment:
Aren't we missing `totalCount` here for the list APIs for adapters, certs, and objects?

Seeing that this was modeled after `GetKeys` in the same SDK, the reason it wasn't included in there is because there exists another API call to count total number of keys in the instance.

Reply:
Done.